File: solutions/csharp/annalyns-infiltration/1/AnnalynsInfiltration.cs
internal static class QuestLogic
{
	public static bool CanFastAttack(bool isKnightAwake) =>
		!isKnightAwake;

	public static bool CanSpy(bool isKnightAwake, bool isArcherAwake, bool isPrisonerAwake) =>
		isKnightAwake
		|| isArcherAwake
		|| isPrisonerAwake;

	public static bool CanSignalPrisoner(bool isArcherAwake, bool isPrisonerAwake) =>
		!isArcherAwake
		&& isPrisonerAwake;

	public static bool CanFreePrisoner(bool isKnightAwake, bool isArcherAwake, bool isPrisonerAwake, bool isPetDogPresent) =>
		!isArcherAwake
		&& (
			isPetDogPresent
			|| (!isPetDogPresent && isPrisonerAwake && !isKnightAwake)
		);
}

File: solutions/java/annalyns-infiltration/1/src/main/java/AnnalynsInfiltration.java
class AnnalynsInfiltration {
    public static boolean canFastAttack(boolean isKnightAwake) {
        return !isKnightAwake;
    }

    public static boolean canSpy(boolean isKnightAwake, boolean isArcherAwake, boolean isPrisonerAwake) {
        return isKnightAwake
        || isArcherAwake
        || isPrisonerAwake;
    }

    public static boolean canSignalPrisoner(boolean isArcherAwake, boolean isPrisonerAwake) {
        return !isArcherAwake
               && isPrisonerAwake;
    }

    public static boolean canFreePrisoner(boolean isKnightAwake, boolean isArcherAwake, boolean isPrisonerAwake, boolean isPetDogPresent) {
        return !isArcherAwake
               && (
                       isPetDogPresent
                       || (isPrisonerAwake && !isKnightAwake)
               );
    }
}
